src/model/operation/transform.js

```diff
@@ -71,47 +112,142 @@ export function transform( a, b, context = {} ) {
+	//
+	// 2. We take next (or first) operation from `operationsA` and check from which operation `b` we need to start transforming it.
+	// All original `operationsA` are set to be transformed starting from the first operation `b`.
+	//
+	// 3. We take operations from `operationsB`, one by one, starting from the correct one, and transform operation `a`
+	// by operation `b` (and vice versa). We update `operationsA` and `operationsB` by replacing the original operations
+	// with the transformation results.
+	//
+	// 4. If operation is broken into multiple operations, we save all the new operations in the place of the
+	// original operation.
+	//
+	// 5. Additionally, if operation `a` was broken, for the "new" operation, we remember from which operation `b` it should
+	// be transformed by.
+	//
+	// 6. We continue transforming "current" operation `a` until it is transformed by all `operationsB`. Then, go to 2.
+	// unless the last operation `a` was transformed.
+	//
+	// The actual implementation of the above algorithm is slightly different, as only one loop (while) is used.
+	// The difference is that we have "current" `a` operation to transform and we store the index of the next `b` operation
+	// to transform by. Each loop operates on two indexes then: index pointing to currently processed `a` operation and
+	// index pointing to next `b` operation. Each loop is just one `a * b` + `b * a` transformation. After each loop
+	// operation `b` index is updated. If all `b` operations were visited for the current `a` operation, we change
+	// current `a` operation index to the next one.
+	//
 
+	// For each operation `a`, keeps information what is the index in `operationsB` from which the transformation should start.
 	const nextTransformIndex = new WeakMap();
 
+	// For all the original `operationsA`, set that they should be transformed starting from the first of `operationsB`.
 	for ( const op of operationsA ) {
 		nextTransformIndex.set( op, 0 );
 	}
 
+	// Additional data that is used for some postprocessing after the main transformation process is done.
 	const data = {
 		nextBaseVersionA: operationsA[ operationsA.length - 1 ].baseVersion + 1,
 		nextBaseVersionB: operationsB[ operationsB.length - 1 ].baseVersion + 1,
-		extraOpsA: 0,
-		extraOpsB: 0
+		originalOperationsACount: operationsA.length,
+		originalOperationsBCount: operationsB.length
 	};
 
+	// Initialize context to use during transformations.
 	const context = initializeContext( operationsA, operationsB, options );
 
+	// Index of currently transformed operation `a`.
 	let i = 0;
 
+	// While not all `operationsA` are transformed...
 	while ( i < operationsA.length ) {
+		// Get "current" operation `a`.
 		const opA = operationsA[ i ];
 
+		// For the "current" operation `a`, get the index of the next operation `b` to transform by.
 		const transformByIndex = nextTransformIndex.get( opA );
```

Comment:
I am not a big fan of this method. `transformByIndex` sounds more like a function than property.